### PR TITLE
Try delaying sibling inserter hover events

### LINKED ISSUE #16646: Spacer block: Can't grab the block's height adjuster
**Describe the bug**
I can't grab the block's height adjuster due to the block below showing the Insert block appender over it.

**To reproduce**
1. Add a Spacer block.
2. Add another block below it.
3. Select the Spacer block again.
4. Try moving the cursor to the node on the bottom of the Spacer block to adjust its height.
5. Observe the Block Inserter appender appearing preventing you from grabbing the node.

**Expected behavior**
I expected to be able to grab the node to the adjuster without it being hidden under the block appender.

**Screenshots**

spacer-hover

**Desktop (please complete the following information):**
 - OSX
 - Firefox 68.0
 - Gutenberg 6.1

### PR DESCRIPTION
**Pull request by kjellr**
Closes #16646. 

This picks up on an idea @mapk had: To have the sibling inserter display after a short delay. This allows for users to select controls that would otherwise be blocked by the inserter. For example, in the GIF below you'll see me resizing the spacer block. But if I stop and wait for `0.75s`, it'll bring up the sibling inserter instead. 

mover

In #17136, @mapk noted some issues around implementing this: 

> I wanted to try a delay on the Inserter, but animating the display attribute isn't possible with just CSS. I think that's why the Inserter uses the opacity attribute instead, but opacity still takes up space even when opacity is set to 0. This results in an element that still covers other UI elements.

Yeah. 😞 Also, we can't set `display: none` or `visibility: hidden` here because we want the sibling inserter to be selectable by keyboard users. That's the reason why it's technically always visible today. 

To get around all that, I adopted a sort of weird approach: 

- I applied a delayed fade animation to the inserter parent. 
- I set the button itself to have `pointer-events: none`, so it ignores clicks when it's hidden, but is still focusable via the keyboard. 
- I added a second delayed animation to the button itself, which just activates `pointer-events: auto` after the same delay used for the parent's animation. 

A couple caveats: 

- I've got this working in Chrome + FF on my end, but Safari won't let me click on the button after it fades into view.
- I haven't tested in IE. It _should_ work there (or at least, fall back to something like its prior behavior). 

This is most definitely a hack approach, and I think we could probably come up with something better that uses javascript? But I figured I'd share anyway in case it helps nudge us along, or in case we're able to sort out the issues it has in Safari.

### CODE REVIEW

**Comment by jasmussen** [comment]
Love this, so good:

delay

This feels like a VASTLY better experience. Not just because it solves the problem of the resizer, but because it also makes the UI less jumpy and heavy. Two upsides that in my opinion strongly outweigh the theoretical extra time you have to wait if you only want to insert a block using the sibling inserter.

Now you got me looking though, and I noticed this:

Screenshot 2019-08-29 at 09 12 23

☝️ I used the inspector to make sure the on-select borders were always visible for all blocks. And in doing so it's clear that the sibling inserter isn't centered between two blocks. Can it be?

File: packages/block-editor/src/components/block-list/style.scss [hunk 1]
```diff
@@ -879,7 +880,17 @@
 
 	&:hover,
 	&.is-visible {
-		opacity: 1;
+		// Only show this inserter after a delay.
+		animation: block-editor-block-list__insertion-point-inserter-fade-in 0.2s ease-out 0.75s;
```
[the remark this answers is not in the document]
Comment (jasmussen) [reply packages/block-editor/src/components/block-list/style.scss]: Does the "reduce motion" thing suggest either approach?

**Comment by kjellr** [comment]
Fixed in https://github.com/WordPress/gutenberg/pull/17240/commits/7ee67d7c661004106a030ec74b0c4d51fd71769a. 

Screen Shot 2019-08-29 at 9 51 32 AM

**Comment by kjellr** [comment]
I went down a rabbit hole with this one yesterday afternoon, and hope to spend a little time on it today with fresh eyes.

A general update:

- The delayed hover appearance/functionality of the sibling inserter seems to be well received above.
- As per [discussions above](https://github.com/WordPress/gutenberg/pull/17240#discussion_r319118483), CSS-only may not be the best route for this one. I have a mostly working JS version that I'll push up if I can sort out the bugs. 
- Either way, we'll certainly make sure that the button is focusable via the keyboard, and appears immediately when it is focused.

**Comment by kjellr** [comment]
Alright. I've had lots of time to dig in, and even though I _thought_ I was closer, I hit a wall and it turns out I'm even farther from a solution than originally thought. 😅 

These inserters have 4 levels: 

1. First, there's a parent `div`, named `editor-block-list__insertion-point`
2. Then there's a child `div`, called `block-editor-block-list__insertion-point-inserter`
3. Then there's a child of that called `block-editor-inserter`
4. Finally, inside of that of that is the button, called `block-editor-inserter__toggle`

All of these levels slightly overlap the preceding block. So in this screenshot, the insertion point area that's highlighted belongs to the paragraph block:

Screen Shot 2019-08-30 at 9 39 58 AM

That's what we all end up clicking when we reach for the movers. Even if we were to rip out all of the levels below that, the top parent would still be blocking our clicks. 

I'd been operating under the impression that the final level (the button itself) was the element that was intercepting hovers + clicks. But that's not the case! _All_ of those levels sit on top of the preceding block, so _all_ of them block hovers/clicks. 

We'd actually need to disable pointer events on that top parent div (and all children) in order to still use the resize controller like normal. But that won't work either: if we turn off pointer events on the parent, then we'll have nothing hover on, which means we'll never end up showing the toggle button in the first place. I haven't yet found a decent solution (css or js) that allows us to let clicks through, but will still trigger the hovers. I've tried some workarounds, but they usually result in two competing hover states firing off, which isn't great:

uh-oh

I'm open to solutions to that if anyone has them! But in the meantime I think this PR is likely dead. 😞 To solve the overlap issue, I think the only simple solution would be to narrow the height of the inserter area (the one in that screenshot above), so that it interferes less with the resize controller. Unfortunately, that'd mean that mouse users would have to be a bit more fiddly when they want to activate that inserter: they'd end up with a smaller hover area. 

Anyway — if nobody has any good suggestions, I think we're good to close this PR in the meantime.

**Comment by kjellr** [comment]
cc @youknowriad or @ellatrix since you two seem to have lots of creative solutions to these sorts of things. 🙂

In general: we'd like to stop the sibling inserter from blocking clicks, but still keep it hoverable in some way — so that if you were to leave your mouse in between blocks, we would still surface the inserter after a little delay.

**Comment by ZebulanStanphill** [comment]
#17136 just got merged, which sort of "fixes" the issue where you can't reach the drag handle.

Ultimately, I think the sibling inserter should never overlap the block that you are editing. It should appear outside of the block's borders. I made #13571, which suggests an alternative design.

**Comment by mapk** [comment]
> Anyway — if nobody has any good suggestions, I think we're good to close this PR in the meantime.

So frustrating! Sorry about that, Kjell. 

@ZebulanStanphill has an interesting solution that I think may be worth looking into. My favorite part is moving the sibling inserter _below_ the block rather than above it. Could that help resolve some of the overlap? I realize it would overlap in a different way, but it might be in a more intuitive way.

**Comment by paaljoachim** [comment]
A spontaneous thought.:)
What if the spacer had some bottom margin built into the backend block so that the resizer would not be covered up by the inserter for the next block?

**Comment by kjellr** [comment]
This one seems blocked for now — I'm going to close, and we can open a new PR if we find a JS-based solution that works for us.

**Pull request closed without merging**